### PR TITLE
Bootstrap an automation tool

### PR DESCRIPTION
**Pull request by youknowriad**
We start to have a lot of manual tasks when performing plugin releases, npm releases, changelogs,... and all of these flows. I think we should invest some time automating these to avoid mistakes and I think it could be a huge gain in time ultimately.

To do so, I'd like us to change the "manual" required work from *performing the task* to *checking that everything went well and confirming* using an interactive CLI tool. This is obviously not an easy tool to build with all these flows but I'm confident it will bring more confidence and time gains.

I started looking at the plugin RC release process and bootstrapped a `release-plugin-rc` CLI tool. This flow is complex, this PR implements the following flow:

 - Checking the status of the branch and local changes.
 - Choosing the right version for the RC.
 - Creating the release branch.
 - Performing the bump (updated the files).
 - Committing the version bump changes. 
 - Building the plugin zip. 
 - Pushing the release branch and the git tag.
 - Creating a github release with the changelog and the zip.
 - Cherry-picking the version bump to master.

The process is covered entirely, we could push it a bit further and include slack notifications. 

This tool uses under the hood some npm utilities (commander, inquirer, chalk, simplegit), this adds a small burden when performing `npm install` on the project so I considered whether this should be a separate `package.json`/tool independent from the repository that you can bootstrap when needed. (This could be an easy follow-up if ever this grows too much).

Thoughts?

**Testing instructions**

The best way to try this locally without touching the official repo is to:

 1- Checkout this branch.
 2- Change the value of the `repoOwner` variable (set your username instead of WordPress).
 3- have a fork the Gutenberg repository.
 4- Run the command `bin/commander.js rc`.

At that point, your forked repo should be updated with the release.

### CODE REVIEW

File: bin/commander.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+const SimpleGit = require( 'simple-git/promise' );
+
+// Common info
+const rootFolder = path.resolve( __dirname, '../' );
+const packageJsonPath = rootFolder + '/package.json';
+const packageLockPath = rootFolder + '/package-lock.json';
+const pluginFilePath = rootFolder + '/gutenberg.php';
+const packageJson = require( packageJsonPath );
+const packageLock = require( packageLockPath );
+const simpleGit = SimpleGit( rootFolder );
```
Comment (gziolo) [inline bin/commander.js]: I guess `simple-git` needs to be added as dev dependency as well.

**Review by gziolo — COMMENTED** [review]
This is cool. Let's merge it and start using it and refine as we discover issues :)

**Comment by youknowriad** [comment]
I updated this, this covers the whole RC process now 🎉 
I added testing instructions to be able to test this without touching this repo

**Comment by youknowriad** [comment]
I noticed while developping this that it's hard to develop the script on the repo and have the script perform commits, releases.. on the same repo.

In the last commit, I'm using a separate clone of the repository to perform all the release taks. This has a few of advantages:

 - Make it easy to develop and try the tool
 - Avoid conflicts with existing uncommited changes... the repo is kept untouched
 - Makes it easy to extract as a separate tool later.

**Comment by youknowriad** [comment]
Anyone willing to give this a try before I move forward and merge it. It should be easy to test on a fork (see testing instructions)

File: bin/commander.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+				tag_name: 'v' + nextVersion,
+				name: nextVersionLabel,
+				body: changelog,
+				prerelease: true,
+			} );
+			release = releaseData.data;
+
+			console.log( '>> The Github release has been created succesfully.' );
+		} );
+		abortMessage = 'Aborting! Make sure to remove the remote release branch, the git tag and the Github release.';
```
Comment (aduth) [inline bin/commander.js]:
```suggestion
		abortMessage = 'Aborting! Make sure to remove the remote release branch, the git tag and the GitHub release.';
```

File: bin/commander.js [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+				owner: repoOwner,
+				repo: 'gutenberg',
+				tag_name: 'v' + nextVersion,
+				name: nextVersionLabel,
+				body: changelog,
+				prerelease: true,
+			} );
+			release = releaseData.data;
+
+			console.log( '>> The Github release has been created succesfully.' );
```
Comment (aduth) [inline bin/commander.js]:
```suggestion
			console.log( '>> The GitHub release has been created succesfully.' );
```

File: bin/commander.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+			console.log( '>> The plugin zip has been succesfully uploaded.' );
+		} );
+		abortMessage = 'Aborting! Make sure to manually cherry-pick the ' + success( commitHash ) + ' commit to the master branch.';
+
+		// Cherry-picking the bump commit into master
+		await runStep( 'Cherry-picking the bump commit into master', abortMessage, async () => {
+			await askForConfirmationToContinue(
+				'The plugin RC is now released. Proceed with the version bump in the master branch?',
+				true,
+				'test'
```
Comment (aduth) [inline bin/commander.js]: Debugging remnant?

File: bin/commander.js [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+		} );
+
+		// Bumping the version in the different files (package.json, package-lock.json, gutenberg.php)
+		let commitHash;
+		await runStep( 'Updating the plugin version', abortMessage, async () => {
+			const newPackageJson = {
+				...packageJson,
+				version: nextVersion,
+			};
+			fs.writeFileSync( packageJsonPath, JSON.stringify( newPackageJson, null, '\t' ) + '\n' );
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/commander.js]: I wondered the same but in my testing, the order is kept

File: bin/commander.js [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+const os = require( 'os' );
+const uuid = require( 'uuid/v4' );
+
+// Common info
+const workingDirectoryPath = path.join( os.tmpdir(), uuid() );
+const packageJsonPath = workingDirectoryPath + '/package.json';
+const packageLockPath = workingDirectoryPath + '/package-lock.json';
+const pluginFilePath = workingDirectoryPath + '/gutenberg.php';
+const gutenbergZipPath = workingDirectoryPath + '/gutenberg.zip';
+const repoOwner = 'WordPress';
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/commander.js]: I'm hesitant to introduce this now. I think there's value in it but down the road, we're going to touch SVN repos as well. Will we need an env variable for that repo as well. Ideally we'd have a consistent pattern to ease testing in fake environments. Maybe a single config file?

File: bin/commander.js [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+			const { changelog } = await inquirer.prompt( [ {
+				type: 'editor',
+				name: 'changelog',
+				message: 'Please provide the CHANGELOG of the release (markdown)',
+			} ] );
+
+			const { token } = await inquirer.prompt( [ {
+				type: 'input',
+				name: 'token',
+				message: 'Please provide a Github personal authentication token. (Navigate to ' + success( 'https://github.com/settings/tokens' ) + ' to create one with admin:org, repo and write:packages rights)',
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply bin/commander.js]: I don't think we are, but I think for the "roles", GitHub considers the releases as packages. Without that right, I had errors.

**Comment by youknowriad** [comment]
This should be ready, I'm going to merge this later today to be able to use it for the next release.

**Review by gziolo — APPROVED** [review]
It works like a charm. I was able to go through all steps. This is the result of my test run:

https://github.com/gziolo/gutenberg/releases/tag/v5.8.0-rc.1

This is pure gold. I love all the bits of this script. My favorite part is that it operates on fresh clone of the repo so it doesn't remove my project configuration in IDE 🎉

**Comment by gziolo** [comment]
By the way, let's add a note in release notes to make sure people use it :)

**Comment by youknowriad** [comment]
> By the way, let's add a note in release notes to make sure people use it :)

I'll do it once we figure out the whole process.

**Pull request merged**